Pack cache options (#149)
buildpacks/rfcs: text/0000-pack-cache-options.md

# Meta
[meta]: #meta
- Name: Pack cache options
- Start Date: (fill in today's date: YYYY-MM-DD)
- Author(s): [@jromero](https://github.com/jromero), [@dwillist](https://github.com/dwillist)
- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: N/A

# Summary
[summary]: #summary

<!-- One paragraph explanation of the feature. -->

Allow users to configure the various cache options.

# Definitions
[definitions]: #definitions

<!-- Make a list of the definitions that may be useful for those reviewing. Include phrases and words that buildpack authors or other interested parties may not be familiar with. -->

#### General

* **bind mounts** - Directories mounted from the docker host system onto the running container.
* **named volumes** - Volumes that are referenced by name within docker and not directly associated with the host file system.
* **lifecycle** - Component that executes in a container running all the necessary buildpacks and producing the final app image.

#### Cache Types

> * **build cache** - Cache container `build` layers. That is layers that were created by buildpacks and marked as `build=true` and `cache=true`.

Reviewer: ```suggestion
* **build cache** - Cache containing layers that should be restored on subsequent builds. That is layers that were created by buildpacks and marked as `cache=true`.
```
Layers don't need to be `build=true` to be cached.

Author: That's interesting and honestly confusing but I'll take your word for it. :)

- Explaining the feature largely in terms of examples.
- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
- If applicable, describe the differences between teaching this to existing users and new users. -->

A new variadic flag that enables configuring different cache types geared towards app developers or operators that are using pack in very specific (and sometimes restrictive) ways that require them to configure caching in a more advanced fashion.

The options available via this new flag would have sensible defaults based on the current behaiviour with the exception of the `--cache-image` which would be deprecated and later removed to reduce confusion. Overall UX would remain the same while allowing for greater flexibility.

An additional benefit of this format is that it can later be expanded on if more options for a specific format were deamed necessary.

```
pack build --cache 'type=<type>;format=<format>;<additional-options>' --cache ...
```

### `cache` flag options
* _**type**_ - The type of cache to configure [`build`, `launch`] (default: `build`).

> * _**format**_ - The format in which to store that cache [`dir`, `image`, `volume`] (default: `volume`). 

Reviewer: I am hesitant to expose the launch cache concept to users. It is a hack necessary to produce adequate performance when exporting to a docker daemon. Instead f exposing this concept to users I would rather drop it entirely if we stop supporting the daemon as an export target (as we have previously discussed).

Making it directly configurable only opens up potential confusion imho, for example if we used this cache during a registry build it would actually decrease performance. Also, neither layer contents or layer metadata are restored from this cache, making it confusing to buildpack authors concerned with the layer reuse lifecycle.

Reviewer: Do we have existing use cases for configuring the launch cache? Could we solve this problem another way?

Reviewer 2: Think [the linked ticket](https://community.atlassian.com/t5/Bitbucket-discussions/Using-CNCF-Buildpacks-on-Bitbucket-Pipelines/td-p/1516690) would require the launch cache to be configurable. We could potentially work around this by making the following changes in pack:
- Explicitly create a launch cache volume before running any of the lifecycle phases.
  - If this errors out because of permissions just ignore it and continue on without a launch cache
  - If it works yay mount it into the lifecycle containers!
- Force people to use image caches in these environments (this is already an ask [here](https://github.com/buildpacks/pack/issues/1024))

Author: As @dwillist mentions, we _need_ to expose users to the launch cache in order to be able to configure it for very specific use cases. Another example just came up in pack-orb (https://github.com/buildpacks/pack-orb/issues/25#issuecomment-824086257).

> The reason for the current requirement appears to be on trying to cache the docker volumes created. Once we have a way to configure cache volumes via pack, we should be able to remove that dependency.

I'm on the boat for removing it and IMO that is one of the benefits of this proposal. In the future, the `type` option would likely not exist.

###### Format specific options

> * `dir`

Reviewer: Should we call this type `bind` instead? And use `source` instead of `path`. If I am familiar with the `--volume` flag syntax, using something slightly different here feels like it would be hard to remember.

If we ever wanted to make the in-container path configurable we could add a `target` option. `path` might be ambiguous if we are setting host and container paths, or a host image and a container path.

Author: I'm okay with this. I opted for `dir` and `path` to make it the most understandable for certain users. My thought process involved the idea that we shouldn't expose the internals (where the cache dir would be mounted) but that may be wrong long-term.

Additionally, I see the value of aligning it with pre-existing concepts.

pack build --cache 'format=image;name=io.docker.io/myorg/my-cache:build'
```

When no flags are provided the following equate to the defaults:

```
# default when no cache flag is provided
pack build --cache 'type=build;format=volume;name=pack-<type>-<app-uuid>' --cache 'type=volume;format=volume;name=pack-<type>-<app-uuid>'
```

> NOTE: Templating as done internally by Pack for `<type>` and `<app-uuid>` is not a proposed part of this solution.

# How it Works
[how-it-works]: #how-it-works

<!-- This is the technical portion of the RFC, where you explain the design in sufficient detail.

The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work. -->

Pack would parse the `--cache` flags and apply any options to the two particular cache types. It would take the necessary steps to provide the approprite configuration when executing the lifecycle.

# Drawbacks
[drawbacks]: #drawbacks
<!-- 
Why should we *not* do this? -->

#### Pros

- The fomat is extensible to allow for additional cache formats or additional options.

#### Cons

- There is added complexity in the syntax since it's freeform.

>     - Given that the target use cases are more advanced this should be too impactful.

Reviewer: Not saying it isn't worth it, but I don't think we should minimize the impact of creating complex custom usage. Ever since we introduce the buildpack URIs I find myself checking the docs frequently because I have trouble remembering the syntax, and I would consider myself an "advanced" user.

Author: Don't disagree but would you rather the alternative be used or is there another proposed option?

Reviewer: Could we do a hybrid of this proposal and the listed alternative and remove `type` from this syntax. e.g. provide `--build-cache` and `--launch-cache` (or `--layers-cache` and `--daemon-cache`?) flags? I think this would allow `pack` to provide more meaningful help text for each type of cache and would allow us to more explicitly remove concepts like the launch cache, if we ever wanted to do that.

Author: @dwillist / @dfreilich what do y'all think about this proposal? I'm okay with `--build-cache` and `--launch-cache` and agree with the sentiment.

##### Examples

```bash
# specify named volume build cache
pack build --build-cache-volume my-build-cache

# specify bind mount cache (relative to current working directory)
pack build --build-cache-dir ./my-build-cache

# specify named volume build cache (name only)
pack build --build-cache-volume my-cool-build-cache

# specify image cache
pack build --build-cache-image io.docker.io/myorg/my-cache:build
```

When no flags are provided the following equate to the defaults:

```
# default when no cache flag is provided
pack build --build-cache-volume my-app-build-cache --launch-cache-volume my-app-build-cache
```

##### Pros

- Flags are more human readable and minimal.

##### Cons

- Flags need to be treated as exclusive which may add confusion. 
    
    For example the following should be invalid:
    ```
    pack build --build-cache-image <image> --build-cache-volume <volume>
    ``` 
- The number of flags pollute the output of `pack build --help`.
- If additional options are necessary, a new flag must be introduced.

# Prior Art
[prior-art]: #prior-art
<!-- 
Discuss prior art, both the good and bad. -->

> * [Docker CLI `--mount` flag](https://docs.docker.com/storage/bind-mounts/):
>     `--mount type=bind,source="$(pwd)"/target,target=/app`
> 

Author: ```suggestion
* [Docker CLI `--mount` flag](https://docs.docker.com/storage/bind-mounts/):
    `--mount type=bind,source="$(pwd)"/target,target=/app`
* [BuildKit `--output` flag](https://github.com/moby/buildkit#output):
    `--output type=image,name=docker.io/username/image,push=true`
```

- If applicable, describe the differences between teaching this to existing users and new users. -->

A new variadic flag that enables configuring different cache types geared towards app developers or operators that are using pack in very specific (and sometimes restrictive) ways that require them to configure caching in a more advanced fashion.

The options available via this new flag would have sensible defaults based on the current behaiviour with the exception of the `--cache-image` which would be deprecated and later removed to reduce confusion. Overall UX would remain the same while allowing for greater flexibility.

An additional benefit of this format is that it can later be expanded on if more options for a specific format were deamed necessary.

```
pack build --cache 'type=<type>;format=<format>;<additional-options>' --cache ...
```

### `cache` flag options
* _**type**_ - The type of cache to configure [`build`, `launch`] (default: `build`).
* _**format**_ - The format in which to store that cache [`bind`, `image`, `volume`] (default: `volume`). 

###### Format specific options

* `bind`
    * _**source**_ - Path to the local directory to use as cache.

> * `image`

Reviewer 3: And, to clarify (per https://github.com/buildpacks/pack/issues/1024#issuecomment-762901676), this image could be a local image, or would it still need to be a registry image?

Author: The linked comment does bring up the _future_ potential for an additional flag similar to [buildkit](https://github.com/moby/buildkit#output):

```
push=<boolean>
```

ie:

```
type=build;format=image;name=io.docker.io/myorg/my-cache:build;push=false
```

    * _**name**_ - Name of the image to use to save as cache (including tag).
* `volume`
    * _**name**_ - Name of the volume to use to save as cache. (default: `pack-<type>-<app-uuid>`)

##### Examples

```bash
# specify named volume build cache
pack build --cache 'type=build;format=volume;name=my-build-cache'

# specify bind mount cache (relative to current working directory)

> pack build --cache 'type=build;format=bind;source=./my-build-cache'

Reviewer 3: ```suggestion
pack build --cache 'type=build;format=bind;name=./my-build-cache'
```
Where possible, I'd like to keep the same option names for all options (if that's sensible), to minimize the loss factor when switching cache types.

Author: I can't seem to find @ekcasey's comment that inspired the change from `format=dir;path=./my-build-cache` to match Docker's [`mount`](https://docs.docker.com/storage/bind-mounts/#choose-the--v-or---mount-flag) format.

Personally, I think it's more confusing to use `name` because it doesn't clearly depict what "name" means in this context whereas `source` is pretty clear, especially if you are familiar with `mount`.
